### PR TITLE
[Doc] Reorg API Documentation part-1

### PR DESCRIPTION
**Pull request by zhjwy9343**
*Issue #, if available:*

*Description of changes:*
This PR reorganized GS API documentation in the following perspectives:
    - Put all Base classes into their corresponding module;
    - Add docstring for the graphstorm module;
    - Create individual template.rst for each module so that they only display their specific APIs;
    - Fix the .. code:: directive error in the GSgnnModelBase class.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/source/api/graphstorm.eval.rst [hunk 1]
```diff
@@ -0,0 +1,38 @@
+.. _apieval:
+
+graphstorm.eval
+=======================
+
+    GraphStorm provides built-in evaluation methods for different Graph Machine
+    Learning (GML).
```
Comment (classicsong) [inline docs/source/api/graphstorm.eval.rst:7]:
```suggestion
    Learning (GML) tasks.
```

File: docs/source/api/graphstorm.inference.rst [hunk 2]
```diff
@@ -13,7 +13,7 @@ graphstorm.inferrer
 .. autosummary::
     :toctree: ../generated/
     :nosignatures:
-    :template: classtemplate.rst
+    :template: pipelinetemplate.rst
```
Comment (classicsong) [inline docs/source/api/graphstorm.inference.rst:16]: GSInferrer does not have save_model, remove_saved_model, save_topk_models, get_best_model_path, restore_model, etc.

File: docs/source/api/graphstorm.rst [hunk 3]
```diff
@@ -9,13 +9,14 @@ graphstorm
     Users can directly use the following code to use these functions.
 
     >>> import graphstorm as gs
-    >>> gs.initialize()
-    >>> gs.get_rank()
+    >>> gs.initialize(ip_config="/tmp/ip_list.txt", backend="gloo")
+    >>> gs.setup_device(local_rank)
+    >>> gs.get_feat_size(g, "ntyp1")
```
Comment (classicsong) [inline docs/source/api/graphstorm.rst:14]: Why users need to call gs.get_feat_size(g, "ntyp1") ?

Only the first two are recommended.

File: python/graphstorm/gsf.py [hunk 4]
```diff
@@ -111,11 +111,12 @@ def get_feat_size(g, node_feat_names):
     g : DistGraph
         The distributed graph.
     node_feat_names : str or dict of str
-        The node feature names.
+        The node type names for extracting feature size.
```
Comment (classicsong) [inline python/graphstorm/gsf.py:114]: This is node feature names, actually.

File: python/graphstorm/model/edge_gnn.py [hunk 5]
```diff
@@ -89,8 +89,10 @@ def predict(self, blocks, target_edges, node_feats, edge_feats,
             is true otherwise return the maximum value.
         """
 
-class GSgnnEdgeModelBase(GSgnnModelBase,  # pylint: disable=abstract-method
-                         GSgnnEdgeModelInterface):
+# pylint: disable=abstract-method
+class GSgnnEdgeModelBase(GSgnnEdgeModelInterface,
+                         GSgnnModelBase
```
Comment (classicsong) [inline python/graphstorm/model/edge_gnn.py:94]:
```suggestion
                         GSgnnModelBase):
```

File: python/graphstorm/utils.py [hunk 6]
```diff
@@ -30,7 +30,17 @@
...
+    
+    Parameters
+    -----------
+    local_rank: int
+        Rank of the current process in a distributed environment.
+
+    Returns
+    -------
+    str: device where the model runs.
+
```
Comment (classicsong) [inline python/graphstorm/utils.py:43]: remove the empty line.

File: docs/source/api/graphstorm.dataloading.rst [hunk 7]
```diff
@@ -3,29 +3,46 @@
 graphstorm.dataloading
 ==========================
 
-    GraphStorm dataloading module includes a set of graph datasets and dataloaders for different
+    GraphStorm dataloading module includes a set of graph DataSets and DataLoaders for different
     graph machine learning tasks.
 
+    If users would like to customize DataLoaders, please extend those ``***DataLoaderBase`` classes
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply docs/source/api/graphstorm.dataloading.rst:9]: changed wording.

File: docs/source/api/graphstorm.dataloading.rst [hunk 8]
```diff
@@ -3,29 +3,46 @@
...
+
 .. currentmodule:: graphstorm.dataloading
 
+Base DataLoaders
+-------------------
+
+.. autosummary::
+    :toctree: ../generated/
+    :nosignatures:
+    :template: datatemplate.rst
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply docs/source/api/graphstorm.dataloading.rst:20]: The datatemplate include both dataloader and dataset method.

File: docs/source/api/graphstorm.inference.rst (hunk 2 above)
Comment (zhjwy9343) [reply docs/source/api/graphstorm.inference.rst:16]: the pipelinetemplate is share by both trainers and inferrers. If there are some methods missing, the Sphinx complior will raise Warnings, but just ignore those missing methods.

File: docs/source/api/graphstorm.rst (hunk 3 above)
Comment (zhjwy9343) [reply docs/source/api/graphstorm.rst:14]: Some old codes call it. No need anymore. Removed.

File: python/graphstorm/gsf.py (hunk 4 above)
Comment (zhjwy9343) [reply python/graphstorm/gsf.py:114]: got it. reversed.

File: python/graphstorm/utils.py (hunk 6 above)
Comment (zhjwy9343) [reply python/graphstorm/utils.py]: done

File: python/graphstorm/model/lp_gnn.py [hunk 9]
```diff
@@ -54,8 +54,10 @@ def forward(self, blocks, pos_graph, neg_graph,
         The loss of prediction.
         """
 
-class GSgnnLinkPredictionModelBase(GSgnnModelBase,  # pylint: disable=abstract-method
-                                   GSgnnLinkPredictionModelInterface):
+# pylint: disable=abstract-method
+class GSgnnLinkPredictionModelBase(GSgnnLinkPredictionModelInterface,
+                                   GSgnnModelBase,
```
Comment (classicsong) [inline python/graphstorm/model/lp_gnn.py:59]:
```suggestion
                                   GSgnnModelBase):
```

File: python/graphstorm/model/node_gnn.py [hunk 10]
```diff
@@ -80,8 +80,10 @@ def predict(self, blocks, node_feats, edge_feats, input_nodes, return_proba):
             The GNN embeddings.
         """
 
-class GSgnnNodeModelBase(GSgnnModelBase,  # pylint: disable=abstract-method
-                         GSgnnNodeModelInterface):
+# pylint: disable=abstract-method
+class GSgnnNodeModelBase(GSgnnNodeModelInterface,
+                         GSgnnModelBase
```
Comment (classicsong) [inline python/graphstorm/model/node_gnn.py:85]: ditto

**Review by classicsong — APPROVED** [review]

File: python/graphstorm/model/lp_gnn.py (hunk 9 above)
Comment (zhjwy9343) [reply python/graphstorm/model/lp_gnn.py]: changed.

File: python/graphstorm/model/node_gnn.py (hunk 10 above)
Comment (zhjwy9343) [reply python/graphstorm/model/node_gnn.py]: changed

**Pull request merged**